examples/bone_finetuning/finetune.py

```diff
@@ -0,0 +1,371 @@
+import copy
+import logging
+import os
+import random
+from dataclasses import dataclass, field
+from typing import Dict, List, Literal, Optional, Sequence
+
+import datasets
+import numpy as np
+import torch
+import torch.distributed
+import transformers
+from datasets import load_dataset
+from transformers import BitsAndBytesConfig, Trainer
+from transformers.trainer_utils import PREFIX_CHECKPOINT_DIR
+
+from peft import BoneConfig, LoraConfig, PeftModel, TaskType, get_peft_model, prepare_model_for_kbit_training
+
+
+IGNORE_INDEX = -100
+EOT_TOKEN = "<|EOT|>"
+logger = logging.getLogger(__name__)
+
+PROMPT = (
+    "Below is an instruction that describes a task. "
+    "Write a response that appropriately completes the request.\n\n"
+    "### Instruction:\n{instruction}\n\n### Response:"
+)
+
+
+@dataclass
+class TrainingArguments(transformers.TrainingArguments):
+    # Base model or residual model setting
+    model_name_or_path: Optional[str] = field(default="meta-llama/Meta-Llama-3-8B")
+    attn_implementation: Optional[str] = field(default="flash_attention_2")
+    # Lora or PiSSA setting
+    use_lora: Optional[bool] = field(default=False)
+    adapter_name_or_path: Optional[str] = field(
+        default=None,
+        metadata={
+            "help": (
+                "Pre-initialized PiSSA adapter path; when this is not None, the following arguments are ignored."
+            ),
+        },
+    )
+    init_lora_weights: Literal[True, "pissa_niter_4", "olora"] = field(
+        default=True,
+        metadata={
+            "help": ("True -> LoRA; `pissa` -> PiSSA"),
+        },
+    )
+    target_modules: Optional[str] = field(default="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj")
+    lora_rank: Optional[int] = field(default=8)
+    lora_alpha: Optional[float] = field(default=32.0)
+    lora_dropout: Optional[float] = field(
+        default=0.0,
+        metadata={
+            "help": ("Must be set to 0 when using PiSSA."),
```

Comment:
There is a lot of mentioning of PiSSA here but I don't see the relation to Bone. Is this maybe a copy-paste error?

Reply:
The bone was at the bottom, now I have moved it above the lora setting.